testcontainers/src/core/client.rs

```diff
@@ -276,6 +291,66 @@ impl Client {
             .map_err(ClientError::StartContainer)
     }
 
+    pub(crate) async fn copy_to_container(
+        &self,
+        container_id: impl Into<String>,
+        target: String,
+        source: CopyDataSource,
+    ) -> Result<(), ClientError> {
+        let container_id: String = container_id.into();
+
+        let target_directory = match Path::new(&target).parent() {
+            Some(v) => v.display().to_string(),
+            None => return Err(ClientError::PathNameError(target)),
+        };
+
+        let options = UploadToContainerOptions {
+            path: target_directory,
+            no_overwrite_dir_non_dir: "false".into(),
+        };
+
+        let tar = Client::tar_data(target, source).await?;
+
+        self.bollard
+            .upload_to_container::<String>(&container_id, Some(options), tar)
+            .await
+            .map_err(ClientError::CopyToContainer)
+    }
+
+    async fn tar_data(target: String, source: CopyDataSource) -> Result<Bytes, ClientError> {
+        let mut ar = tokio_tar::Builder::new(Vec::new());
+        match source {
+            CopyDataSource::File(file_path) => {
+                let mut f = &mut tokio::fs::File::open(file_path)
+                    .await
+                    .map_err(ClientError::TarError)?;
+                ar.append_file(target, &mut f)
+                    .await
+                    .map_err(ClientError::TarError)?;
+            }
+            CopyDataSource::Data(data) => {
+                let path = Path::new(&target);
+                let file_name = match path.file_name() {
+                    Some(v) => v,
+                    None => return Err(ClientError::PathNameError(target)),
+                };
+
+                let mut header = tokio_tar::Header::new_gnu();
+                header.set_size(data.len() as u64);
+                header.set_mode(0o0644);
+                header.set_cksum();
+
+                ar.append_data(&mut header, file_name, data.as_slice())
+                    .await
+                    .map_err(ClientError::TarError)?;
+            }
+        }
+
+        let bytes = ar.into_inner().await.map_err(ClientError::TarError)?;
+
+        Ok(Bytes::copy_from_slice(bytes.as_slice()))
+    }
```

Comment:
we actually can move this impl into `CopyDataSource`. It doesn't use the client and separate functionality in general

So it will be quite clear call:

`source.tar(target)`

WDYT?

Reply:
That is a fantastic idea. I moved that function and also half of the parts of the other function to the `copy.rs` file. That feels now way better. Thx